torch/distributions/lowrank_multivariate_normal.py

```diff
@@ -0,0 +1,181 @@
+import math
+
+import torch
+from torch.distributions import constraints
+from torch.distributions.distribution import Distribution
+from torch.distributions.multivariate_normal import (_batch_diag, _batch_mahalanobis, _batch_mv,
+                                                     _batch_potrf_lower, _batch_trtrs_lower,
+                                                     _get_batch_shape)
+from torch.distributions.utils import lazy_property
+
+
+def _batch_vector_diag(bvec):
+    """
+    Returns the diagonal matrices of a batch of vectors.
+    """
+    n = bvec.size(-1)
+    flat_bvec = bvec.reshape(-1, n)
+    flat_bmat = torch.stack([v.diag() for v in flat_bvec])
+    return flat_bmat.reshape(bvec.shape + (n,))
+
+
+def _batch_capacitance_tril(W, D):
+    r"""
+    Computes Cholesky of :math:`I + W.T @ inv(D) @ W` for a batch of matrices :math:`W`
+    and a batch of vectors :math:`D`.
+    """
+    m = W.size(-1)
+    Wt_Dinv = W.transpose(-1, -2) / D.unsqueeze(-2)
+    K = torch.matmul(Wt_Dinv, W).contiguous()
+    K.view(-1, m * m)[:, ::m + 1] += 1  # add identity matrix to K
+    return _batch_potrf_lower(K)
+
+
+def _batch_lowrank_logdet(W, D, capacitance_tril=None):
+    r"""
+    Uses "matrix determinant lemma"::
+        log|W @ W.T + D| = log|C| + log|D|,
+    where :math:`C` is the capacitance matrix :math:`I + W.T @ inv(D) @ W`, to compute
+    the log determinant.
+    """
+    if capacitance_tril is None:
+        capacitance_tril = _batch_capacitance_tril(W, D)
```

Comment:
nit: I would remove the `None` default to make it clear that there is no duplicated computation.